core/include/traccc/edm/spacepoint_collection.hpp

```diff
@@ -39,30 +39,47 @@ class spacepoint : public BASE {
     /// @name Spacepoint Information
     /// @{
 
+    /// Invalid measurement index
+    static constexpr unsigned int INVALID_MEASUREMENT_INDEX =
+        std::numeric_limits<unsigned int>::max();
+
     /// The index of the measurement producing this spacepoint (non-const)
     ///
     /// @return A (non-const) vector of <tt>unsigned int</tt> values
     ///
     TRACCC_HOST_DEVICE
-    auto& measurement_index() { return BASE::template get<0>(); }
+    auto& measurement_index_1() { return BASE::template get<0>(); }
     /// The index of the measurement producing this spacepoint (const)
     ///
     /// @return A (const) vector of <tt>unsigned int</tt> values
     ///
     TRACCC_HOST_DEVICE
-    const auto& measurement_index() const { return BASE::template get<0>(); }
+    const auto& measurement_index_1() const { return BASE::template get<0>(); }
+
+    /// The index of the measurement producing this spacepoint (non-const)
+    ///
+    /// @return A (non-const) vector of <tt>unsigned int</tt> values
+    ///
+    TRACCC_HOST_DEVICE
+    auto& measurement_index_2() { return BASE::template get<1>(); }
```

Comment:
Why do we need to keep two indices?

Reply:
Strip spacepoints consist of the combination of two 1D measurements

Reply:
In case / once we form spacepoints from strip measurements, we need at least / exactly two measurements to form one spacepoint.

This is something that @benjaminhuth needs for his code. The ML model would give us all the spacepoints belonging to the same track, but that then needs to be translated back into measurements, so that we could perform a final Kálmán fit using all those measurements.

Reply:
Ah yes okay, good.

Reply:
> E.g., the ActsExamples SimSpacepoint stores them in a `boost::containers::static_vector<SourceLink, 2>`.

:thinking: At one point I should really make [vecmem::static_vector](https://github.com/acts-project/vecmem/blob/main/core/include/vecmem/containers/static_vector.hpp) work properly in device code. It could be a good candidate here indeed, similar to what Acts does.

Reply:
Can't we just use `std::array<measurement_index, 2>` or just `measurement_index[2]`?

Let's absolutely not do this with jagged vectors though, that sounds awful in terms of ergonomics!

Reply:
The reason I didn't do a simple array was to avoid the extra 32-bit memory movement. :thinking: Though a static vector would have the same overhead as well. (Actually, even a bit more.)

But other than possible performance reasons, an array could technically be used, yes.

Reply:
One could expose the `2` as an integer template parameter, and keep the code flexible in a way, that e.g., for a fast tracking chain with pixel only seeding there could be zero overhead in terms of memory...

Reply:
Technically that wouldn't make things different from having 2 separate types. The device code needs to use a specific count, so we'd still need to define a separate pixel and strip spacepoint typedef for the two cases.

No, a fixed array in which the second element is either valid or not, may indeed be the best middle ground here. :thinking:

Reply:
Okay, any way is good for me I think.